### PR TITLE
tokenreview: add APIAudiences config to generic API server and augment context

### PR DESCRIPTION
**Pull request by mikedanese**
Broken out of https://github.com/kubernetes/kubernetes/pull/62692

```release-note
NONE
```

@kubernetes/sig-auth-pr-reviews

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 1]
```diff
@@ -50,12 +50,15 @@ func init() {
...
 // is invoked to serve the request.
-func WithAuthentication(handler http.Handler, auth authenticator.Request, failed http.Handler) http.Handler {
+func WithAuthentication(handler http.Handler, auth authenticator.Request, failed http.Handler, apiAuds authenticator.Audiences) http.Handler {
 	if auth == nil {
 		glog.Warningf("Authentication is disabled")
 		return handler
 	}
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if len(apiAuds) > 0 {
+			req = req.WithContext(genericapirequest.WithAudiences(req.Context(), apiAuds))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Once the response includes audience indication, will we verify the response audience matches one of apiAuds if non-empty?

**Review by liggitt — COMMENTED** [review]
One question on the eventual verification in the filter, lgtm otherwise

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go (hunk 1 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I suspect we will want to validate that here, in the webhook authenticator and in TokenReview.

**Comment by mikedanese** [comment]
Why won't you approve me tide?

/approve

**Comment by liggitt** [comment]
self-approval is no more. we work for the bots now.

**Pull request merged**